File: fast_and_slow_pointer/palindrome-linkedlist.js
class Node {
    constructor(value, next = null) {
        this.value = value;
        this.next = next;
    }
}

const is_palindromic_linked_list = function (head) {
    let result = true;
    let l = head;
    let r = head.next || head;
    if (!r.next) return (r.value === l.value);

    let bl = {next: undefined};
    let isOdd = true;

    while (r) {
        if(!r.next) isOdd = false;
        r = r.next?.next;
        [bl, bl.next, l] = [l, bl, l.next];
    }

    (isOdd) ? r = l.next : r = l;

    while (bl && r) {
        if (bl.value !== r.value) result = false;
        [bl, l, l.next] = [bl.next, bl, l];
        r = r.next;
    }

    return result;
}

head = new Node(2);
head.next = new Node(4);
head.next.next = new Node(6);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(2);

console.log(`LinkedList cycle start: ${is_palindromic_linked_list(head)}`); // true
//
head.next.next.next.next.next = new Node(2);
console.log(`LinkedList cycle start: ${is_palindromic_linked_list(head)}`); // false

File: fast_and_slow_pointer/python/palindrome-linkedlist.py
class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


def is_palindromic_linked_list(head):
    slow = fast = head
    left_of_slow = None

    while fast and fast.next:
        fast = fast.next.next
        slow, left_of_slow = reverse_nodes(slow, left_of_slow)

    fast = slow if fast is None else slow.next

    while fast:
        if fast.value != left_of_slow.value:
            return False
        fast = fast.next
        left_of_slow, slow = reverse_nodes(left_of_slow, slow)

    return True


def reverse_nodes(leading: Node, trailing):
    _ = leading
    leading = leading.next
    _.next = trailing
    trailing = _
    return leading, trailing


if __name__ == '__main__':
    head = Node(2)
    head.next = Node(4)
    head.next.next = Node(6)
    head.next.next.next = Node(4)
    head.next.next.next.next = Node(2)
    print(f"Is palindrome: {is_palindromic_linked_list(head)}")

    head.next.next.next.next.next = Node(2)
    print(f"Is palindrome: {is_palindromic_linked_list(head)}")
